File: src/server/applications.py
"""Application repository layer backed by ``.threatforest/applications.json``.

Applications are first-class persistent entities in the v2 UX model. Each
record has a stable ID, a user-chosen name, a fixed on-disk run directory,
and a business-context block that the user fills in at creation and can
edit later. All fields of ``BusinessContext`` are required so every run has
authoritative user input to seed the scanner agent with.

Storage format: a single JSON object keyed by ``app_id`` sitting alongside
the existing ``runs/`` directory. Matches the filesystem-first convention
used by the rest of the server — no database.

Uniqueness rules:
- ``name`` — case-insensitive unique across all applications.
- ``project_path`` — one application per folder (409 on collision).

``app_id`` format is a short ULID-style token (time-ordered, URL safe)
independent of name and path so renaming never invalidates URLs.
"""

from __future__ import annotations

import json
import re
import secrets
import threading
from datetime import datetime, timezone
from pathlib import Path

from server.models import (
    Application,
    ApplicationCreateRequest,
    ApplicationUpdateRequest,
    BusinessContext,
)
from server.registry import get_runs_root, slugify


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------

class ApplicationError(Exception):
    """Base class for application-repository errors."""


class ApplicationNotFoundError(ApplicationError):
    """Raised when an ``app_id`` does not resolve to a stored application."""


class ApplicationNameConflictError(ApplicationError):
    """Raised when a proposed name collides with an existing application."""


class ApplicationPathConflictError(ApplicationError):
    """Raised when a proposed ``project_path`` already belongs to another app."""


# ---------------------------------------------------------------------------
# Repository
# ---------------------------------------------------------------------------

def _get_store_path() -> Path:
    """Location of the applications store, alongside ``runs/``."""
    return get_runs_root().parent / "applications.json"


def _generate_app_id() -> str:
    """Generate a short, time-ordered, URL-safe application ID.

    Format: ``app_<8 hex chars time>_<6 hex chars random>``. Not a true ULID
    but stable, sortable, and opaque enough for the UX.
    """
    ts = int(datetime.now(tz=timezone.utc).timestamp())
    return f"app_{ts:08x}{secrets.token_hex(3)}"


def _normalise_path(project_path: str) -> str:
    """Return an absolute, resolved string form of a project path."""
    return str(Path(project_path).expanduser().resolve())


def _normalise_name(name: str) -> str:
    """Case-insensitive, whitespace-collapsed form used for uniqueness checks."""
    return re.sub(r"\s+", " ", name.strip().lower())


def _derive_run_dir_name(name: str, existing: set[str]) -> str:
    """Pick an on-disk folder name derived from *name*, avoiding collisions."""
    base = slugify(name) or "app"
    candidate = base
    suffix = 2
    while candidate in existing:
        candidate = f"{base}-{suffix}"
        suffix += 1
    return candidate


class ApplicationRepository:
    """Thread-safe repository for the applications store.

    All mutation methods take the internal lock and write the file
    atomically (write-to-tmp + rename).
    """

    def __init__(self, store_path: Path | None = None) -> None:
        self._store_path = store_path or _get_store_path()
        self._lock = threading.Lock()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _load(self) -> dict[str, dict]:
        """Return the raw store as a dict keyed by ``app_id``."""
        if not self._store_path.is_file():
            return {}
        try:
            data = json.loads(self._store_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return {}
        if not isinstance(data, dict):
            return {}
        return data

    def _save(self, data: dict[str, dict]) -> None:
        """Atomically write the store."""
        self._store_path.parent.mkdir(parents=True, exist_ok=True)
        tmp = self._store_path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(data, indent=2, sort_keys=True), encoding="utf-8")
        tmp.replace(self._store_path)

    @staticmethod
    def _to_model(record: dict) -> Application:
        return Application.model_validate(record)

    @staticmethod
    def _to_record(app: Application) -> dict:
        return app.model_dump()

    # ------------------------------------------------------------------
    # Public read API
    # ------------------------------------------------------------------

    def list_applications(self) -> list[Application]:
        """Return all stored applications, ordered by ``created_at`` asc."""
        data = self._load()
        apps = [self._to_model(rec) for rec in data.values()]
        apps.sort(key=lambda a: a.created_at)
        return apps

    def get_application(self, app_id: str) -> Application:
        """Return the application for *app_id* or raise ``ApplicationNotFoundError``."""
        data = self._load()
        record = data.get(app_id)
        if record is None:
            raise ApplicationNotFoundError(f"Unknown application: {app_id}")
        return self._to_model(record)

    def find_by_name(self, name: str) -> Application | None:
        """Return an application with a matching (case-insensitive) name, or None."""
        target = _normalise_name(name)
        for app in self.list_applications():
            if _normalise_name(app.name) == target:
                return app
        return None

    def find_by_project_path(self, project_path: str) -> Application | None:
        """Return an application whose ``project_path`` matches, or None."""
        target = _normalise_path(project_path)
        for app in self.list_applications():
            if _normalise_path(app.project_path) == target:
                return app
        return None

    # ------------------------------------------------------------------
    # Public write API
    # ------------------------------------------------------------------

    def create_application(self, request: ApplicationCreateRequest) -> Application:
        """Create a new application, enforcing name + path uniqueness.

        Raises
        ------
        ApplicationNameConflictError
            If ``name`` collides (case-insensitive) with an existing app.
        ApplicationPathConflictError
            If ``project_path`` is already registered to another app.
        """
        with self._lock:
            data = self._load()
            existing = [self._to_model(rec) for rec in data.values()]

            name_key = _normalise_name(request.name)
            for app in existing:
                if _normalise_name(app.name) == name_key:
                    raise ApplicationNameConflictError(
                        f"An application named '{request.name}' already exists."
                    )

            path_key = _normalise_path(request.project_path)
            for app in existing:
                if _normalise_path(app.project_path) == path_key:
                    raise ApplicationPathConflictError(
                        f"An application is already registered for project path "
                        f"'{path_key}': '{app.name}'."
                    )

            now = datetime.now(tz=timezone.utc).isoformat()
            run_dir_names = {app.run_dir_name for app in existing}
            run_dir_name = _derive_run_dir_name(request.name, run_dir_names)

            app = Application(
                id=_generate_app_id(),
                name=request.name.strip(),
                slug=slugify(request.name),
                project_path=path_key,
                business_context=request.business_context,
                created_at=now,
                updated_at=now,
                run_dir_name=run_dir_name,
            )
            data[app.id] = self._to_record(app)
            self._save(data)
            return app

    def update_application(
        self, app_id: str, request: ApplicationUpdateRequest
    ) -> Application:
        """Apply partial updates to an application.

        Only ``name`` and ``business_context`` are user-editable. A rename
        also regenerates ``slug`` but leaves ``run_dir_name`` untouched so
        existing run artefacts on disk stay where they are.

        Raises
        ------
        ApplicationNotFoundError
            If the app does not exist.
        ApplicationNameConflictError
            If the new name collides with another application.
        """
        with self._lock:
            data = self._load()
            record = data.get(app_id)
            if record is None:
                raise ApplicationNotFoundError(f"Unknown application: {app_id}")

            current = self._to_model(record)

            new_name = current.name
            new_slug = current.slug
            if request.name is not None and request.name.strip() != current.name:
                proposed = request.name.strip()
                proposed_key = _normalise_name(proposed)
                for other_id, other_rec in data.items():
                    if other_id == app_id:
                        continue
                    other = self._to_model(other_rec)
                    if _normalise_name(other.name) == proposed_key:
                        raise ApplicationNameConflictError(
                            f"An application named '{proposed}' already exists."
                        )
                new_name = proposed
                new_slug = slugify(proposed)

            new_context: BusinessContext = (
                request.business_context
                if request.business_context is not None
                else current.business_context
            )

            updated = Application(
                id=current.id,
                name=new_name,
                slug=new_slug,
                project_path=current.project_path,
                business_context=new_context,
                created_at=current.created_at,
                updated_at=datetime.now(tz=timezone.utc).isoformat(),
                run_dir_name=current.run_dir_name,
            )
            data[app_id] = self._to_record(updated)
            self._save(data)
            return updated

    def delete_application(self, app_id: str) -> None:
        """Remove the application record. Does not touch run artefacts on disk.

        Raises
        ------
        ApplicationNotFoundError
            If the app does not exist.
        """
        with self._lock:
            data = self._load()
            if app_id not in data:
                raise ApplicationNotFoundError(f"Unknown application: {app_id}")
            data.pop(app_id)
            self._save(data)


# ---------------------------------------------------------------------------
# Module-level singleton — matches RunManager / Config access pattern used
# elsewhere in the server.
# ---------------------------------------------------------------------------

_repository: ApplicationRepository | None = None


def get_repository() -> ApplicationRepository:
    """Return the process-wide ``ApplicationRepository`` singleton."""
    global _repository
    if _repository is None:
        _repository = ApplicationRepository()
    return _repository

File: ts/packages/server/src/applications.ts
/**
 * Application repository layer backed by `.threatforest/applications.json`.
 * Port of `src/server/applications.py`.
 *
 * Applications are first-class persistent entities in the v2 UX model. Each
 * record has a stable ID, a user-chosen name, a fixed on-disk run directory,
 * and a required business-context block.
 *
 * Storage format: a single JSON object keyed by `app_id` sitting alongside the
 * `runs/` directory. Filesystem-first — no database.
 *
 * Uniqueness rules:
 *   - `name`         — case-insensitive unique across all applications.
 *   - `project_path` — one application per folder (409 on collision).
 *
 * `app_id` is a short ULID-style token (time-ordered, URL safe) independent of
 * name and path so renaming never invalidates URLs.
 */
import { randomBytes } from 'node:crypto';
import { mkdirSync, readFileSync, renameSync, statSync, writeFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { homedir } from 'node:os';
import {
  ApplicationSchema,
  type Application,
  type ApplicationCreateRequest,
  type ApplicationUpdateRequest,
  type BusinessContext,
} from '@threatforest/types';
import { getRunsRoot, slugify } from './registry.js';

// ---------------------------------------------------------------------------
// Typed errors
// ---------------------------------------------------------------------------

/** Base class for application-repository errors. */
export class ApplicationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ApplicationError';
  }
}

/** Raised when an `app_id` does not resolve to a stored application. */
export class ApplicationNotFoundError extends ApplicationError {
  constructor(message: string) {
    super(message);
    this.name = 'ApplicationNotFoundError';
  }
}

/** Raised when a proposed name collides with an existing application. */
export class ApplicationNameConflictError extends ApplicationError {
  constructor(message: string) {
    super(message);
    this.name = 'ApplicationNameConflictError';
  }
}

/** Raised when a proposed `project_path` already belongs to another app. */
export class ApplicationPathConflictError extends ApplicationError {
  constructor(message: string) {
    super(message);
    this.name = 'ApplicationPathConflictError';
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Location of the applications store, alongside `runs/`. */
function getStorePath(): string {
  return join(dirname(getRunsRoot()), 'applications.json');
}

/**
 * Generate a short, time-ordered, URL-safe application ID.
 * Format: `app_<8 hex chars time><6 hex chars random>`.
 */
function generateAppId(): string {
  const ts = Math.floor(Date.now() / 1000);
  const tsHex = (ts >>> 0).toString(16).padStart(8, '0');
  return `app_${tsHex}${randomBytes(3).toString('hex')}`;
}

/** Expand a leading `~` to the user's home directory. */
function expandUser(p: string): string {
  if (p === '~') return homedir();
  if (p.startsWith('~/')) return join(homedir(), p.slice(2));
  return p;
}

/** Absolute, resolved string form of a project path. */
function normalisePath(projectPath: string): string {
  return resolve(expandUser(projectPath));
}

/** Case-insensitive, whitespace-collapsed form used for uniqueness checks. */
function normaliseName(name: string): string {
  return name.trim().toLowerCase().replace(/\s+/g, ' ');
}

/** Pick an on-disk folder name derived from `name`, avoiding collisions. */
function deriveRunDirName(name: string, existing: Set<string>): string {
  const base = slugify(name) || 'app';
  let candidate = base;
  let suffix = 2;
  while (existing.has(candidate)) {
    candidate = `${base}-${suffix}`;
    suffix += 1;
  }
  return candidate;
}

// ---------------------------------------------------------------------------
// Repository
// ---------------------------------------------------------------------------

/**
 * Repository for the applications store. Mutations are serialized through an
 * async mutex and persisted atomically (write-to-tmp + rename), mirroring the
 * Python repository's lock + atomic-write semantics in a single-process,
 * cooperatively-scheduled Node runtime.
 */
export class ApplicationRepository {
  private readonly storePath: string;

  /** Serializes mutating sections — the await-chain analogue of the Python lock. */
  private chain: Promise<void> = Promise.resolve();

  constructor(storePath?: string) {
    this.storePath = storePath ?? getStorePath();
  }

  // ------------------------------------------------------------------
  // Internal helpers
  // ------------------------------------------------------------------

  /** Return the raw store as a record keyed by `app_id`. */
  private load(): Record<string, Record<string, unknown>> {
    let exists = false;
    try {
      exists = statSync(this.storePath).isFile();
    } catch {
      exists = false;
    }
    if (!exists) return {};
    try {
      const data = JSON.parse(readFileSync(this.storePath, 'utf-8'));
      if (typeof data !== 'object' || data === null || Array.isArray(data)) return {};
      return data as Record<string, Record<string, unknown>>;
    } catch {
      return {};
    }
  }

  /** Atomically write the store (keys sorted, 2-space indent). */
  private save(data: Record<string, Record<string, unknown>>): void {
    mkdirSync(dirname(this.storePath), { recursive: true });
    const sorted = sortKeysDeep(data);
    const tmp = `${this.storePath}.tmp`;
    writeFileSync(tmp, JSON.stringify(sorted, null, 2), 'utf-8');
    renameSync(tmp, this.storePath);
  }

  /** Run a mutation under the serializing chain (analogue of `with lock`). */
  private async withLock<T>(fn: () => T): Promise<T> {
    const run = this.chain.then(fn);
    // Keep the chain alive even if `fn` throws, so later mutations still run.
    this.chain = run.then(
      () => undefined,
      () => undefined,
    );
    return run;
  }

  private toModel(record: Record<string, unknown>): Application {
    return ApplicationSchema.parse(record);
  }

  private toRecord(app: Application): Record<string, unknown> {
    return { ...app };
  }

  // ------------------------------------------------------------------
  // Public read API
  // ------------------------------------------------------------------

  /** Return all stored applications, ordered by `created_at` ascending. */
  listApplications(): Application[] {
    const data = this.load();
    const apps = Object.values(data).map((rec) => this.toModel(rec));
    apps.sort((a, b) => (a.created_at < b.created_at ? -1 : a.created_at > b.created_at ? 1 : 0));
    return apps;
  }

  /** Return the application for `appId` or throw `ApplicationNotFoundError`. */
  getApplication(appId: string): Application {
    const data = this.load();
    const record = data[appId];
    if (record === undefined) {
      throw new ApplicationNotFoundError(`Unknown application: ${appId}`);
    }
    return this.toModel(record);
  }

  /** Return an application with a matching (case-insensitive) name, or null. */
  findByName(name: string): Application | null {
    const target = normaliseName(name);
    for (const app of this.listApplications()) {
      if (normaliseName(app.name) === target) return app;
    }
    return null;
  }

  /** Return an application whose `project_path` matches, or null. */
  findByProjectPath(projectPath: string): Application | null {
    const target = normalisePath(projectPath);
    for (const app of this.listApplications()) {
      if (normalisePath(app.project_path) === target) return app;
    }
    return null;
  }

  /** Return an application whose on-disk `run_dir_name` matches, or null. */
  findByRunDirName(runDirName: string): Application | null {
    for (const app of this.listApplications()) {
      if (app.run_dir_name === runDirName) return app;
    }
    return null;
  }

  // ------------------------------------------------------------------
  // Public write API
  // ------------------------------------------------------------------

  /**
   * Create a new application, enforcing name + path uniqueness.
   *
   * @throws ApplicationNameConflictError if `name` collides (case-insensitive).
   * @throws ApplicationPathConflictError if `project_path` is already registered.
   */
  async createApplication(request: ApplicationCreateRequest): Promise<Application> {
    return this.withLock(() => {
      const data = this.load();
      const existing = Object.values(data).map((rec) => this.toModel(rec));

      const nameKey = normaliseName(request.name);
      for (const app of existing) {
        if (normaliseName(app.name) === nameKey) {
          throw new ApplicationNameConflictError(
            `An application named '${request.name}' already exists.`,
          );
        }
      }

      const pathKey = normalisePath(request.project_path);
      for (const app of existing) {
        if (normalisePath(app.project_path) === pathKey) {
          throw new ApplicationPathConflictError(
            `An application is already registered for project path ` +
              `'${pathKey}': '${app.name}'.`,
          );
        }
      }

      const now = new Date().toISOString();
      const runDirNames = new Set(existing.map((a) => a.run_dir_name));
      const runDirName = deriveRunDirName(request.name, runDirNames);

      const app: Application = {
        id: generateAppId(),
        name: request.name.trim(),
        slug: slugify(request.name),
        project_path: pathKey,
        business_context: request.business_context,
        created_at: now,
        updated_at: now,
        run_dir_name: runDirName,
      };
      data[app.id] = this.toRecord(app);
      this.save(data);
      return app;
    });
  }

  /**
   * Apply partial updates to an application.
   *
   * `name`, `business_context`, and `project_path` are user-editable. The
   * route layer enforces the "project_path only editable before first run"
   * rule. A rename regenerates `slug` but leaves `run_dir_name` untouched so
   * existing run artefacts stay where they are.
   *
   * @throws ApplicationNotFoundError if the app does not exist.
   * @throws ApplicationNameConflictError if the new name collides.
   * @throws ApplicationPathConflictError if the new path is already registered.
   */
  async updateApplication(
    appId: string,
    request: ApplicationUpdateRequest,
  ): Promise<Application> {
    return this.withLock(() => {
      const data = this.load();
      const record = data[appId];
      if (record === undefined) {
        throw new ApplicationNotFoundError(`Unknown application: ${appId}`);
      }

      const current = this.toModel(record);

      let newName = current.name;
      let newSlug = current.slug;
      if (request.name !== null && request.name.trim() !== current.name) {
        const proposed = request.name.trim();
        const proposedKey = normaliseName(proposed);
        for (const [otherId, otherRec] of Object.entries(data)) {
          if (otherId === appId) continue;
          const other = this.toModel(otherRec);
          if (normaliseName(other.name) === proposedKey) {
            throw new ApplicationNameConflictError(
              `An application named '${proposed}' already exists.`,
            );
          }
        }
        newName = proposed;
        newSlug = slugify(proposed);
      }

      const newContext: BusinessContext =
        request.business_context !== null ? request.business_context : current.business_context;

      let newProjectPath = current.project_path;
      if (request.project_path !== null) {
        const proposedPath = normalisePath(request.project_path);
        if (proposedPath !== normalisePath(current.project_path)) {
          for (const [otherId, otherRec] of Object.entries(data)) {
            if (otherId === appId) continue;
            const other = this.toModel(otherRec);
            if (normalisePath(other.project_path) === proposedPath) {
              throw new ApplicationPathConflictError(
                `An application is already registered for project path ` +
                  `'${proposedPath}': '${other.name}'.`,
              );
            }
          }
          newProjectPath = proposedPath;
        }
      }

      const updated: Application = {
        id: current.id,
        name: newName,
        slug: newSlug,
        project_path: newProjectPath,
        business_context: newContext,
        created_at: current.created_at,
        updated_at: new Date().toISOString(),
        run_dir_name: current.run_dir_name,
      };
      data[appId] = this.toRecord(updated);
      this.save(data);
      return updated;
    });
  }

  /**
   * Remove the application record. Does not touch run artefacts on disk.
   *
   * @throws ApplicationNotFoundError if the app does not exist.
   */
  async deleteApplication(appId: string): Promise<void> {
    return this.withLock(() => {
      const data = this.load();
      if (!(appId in data)) {
        throw new ApplicationNotFoundError(`Unknown application: ${appId}`);
      }
      delete data[appId];
      this.save(data);
    });
  }
}

/** Recursively sort object keys so JSON output is deterministic (sort_keys=True). */
function sortKeysDeep(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value !== null && typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      out[key] = sortKeysDeep((value as Record<string, unknown>)[key]);
    }
    return out;
  }
  return value;
}

// ---------------------------------------------------------------------------
// Module-level singleton — matches RunManager / Config access pattern.
// ---------------------------------------------------------------------------

let _repository: ApplicationRepository | null = null;

/** Return the process-wide `ApplicationRepository` singleton. */
export function getRepository(): ApplicationRepository {
  if (_repository === null) {
    _repository = new ApplicationRepository();
  }
  return _repository;
}

/**
 * Resolve the absolute `project_path` for an application by its opaque
 * `app_id`, or `null` if the app is unknown. Used by `executor.ts` to seed a
 * run with the correct repository path independent of the request's
 * `project_path` (which v2 callers omit in favour of the app reference).
 */
export function resolveProjectPathForApp(appId: string): string | null {
  try {
    return getRepository().getApplication(appId).project_path;
  } catch {
    return null;
  }
}
